pytensor/tensor/rewriting/basic.py

```diff
@@ -910,6 +910,38 @@ def local_join_make_vector(fgraph, node):
         return [ret]
 
 
+@register_canonicalize
+@node_rewriter([Join])
+def local_join_to_repeat(fgraph, node):
+    """Join(axis, x, x, x, ...) -> repeat(x, n, axis)
+
+    When the same tensor is concatenated multiple times,
+    replace with a single repeat operation which is more efficient.
+
+    Examples
+    --------
+    concatenate([x, x, x], axis=0) -> repeat(x, 3, axis=0)
+    """
+    # Extract axis and the tensors being joined
+    axis, *tensors = node.inputs
+
+    # Need at least 2 tensors to consider optimization
+    if len(tensors) <= 1:
+        return
+
+    # Check if all tensors are identical
+    if not all(t == tensors[0] for t in tensors[1:]):
+        return
+
+    # Replace with repeat operation
+    result = repeat(tensors[0], len(tensors), axis)
```

Comment:
one thing you'll need to handle is that join axis is symbolic but repeat must be a python integer. You should do something like:

```python
if not isinstance(axis, Constant):
    return None  # rewrite only applies to constant axis
axis = axis.data
```